In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd
import numpy as np
df=pd.read_csv("/content/drive/MyDrive/datasets/data_ready_to_process(with buy sell).csv")

In [ ]:
df

horizon (days)  amount  Volatility_Buy  Volatility_sell  Sharpe Ratio  \
0                    2     100        0.383666         0.385748      0.383666   
1                  330   15000        0.322809         0.236350      0.322809   
2                    7    3000        0.238642         0.235491      0.238642   
3                    5   20000        0.428559         0.429340      0.428559   
4                  360   15000        0.194612         0.254011      0.194612   
...                ...     ...             ...              ...           ...   
405253             480   50000        0.241103         0.253235      0.241103   
405254             300   50000        0.410470         0.535448      0.410470   
405255             210    3000        0.316220         0.359625      0.316220   
405256             270     500        0.173693         0.218904      0.173693   
405257              15    2000        0.131422         0.145768      0.131422   

        expected_return (yearly)  inflation  nominal_return  ESG_ranking  \
0                       0.001437       1.96       -0.037225         12.0   
1                       0.170969      -0.20        0.324317         26.3   
2                       0.002824      -0.20       -0.005756         19.8   
3                       0.000094      -0.20       -0.058388         12.9   
4                       0.149979      -0.50        0.003444         27.9   
...                          ...        ...             ...          ...   
405253                  0.053763      -0.15       -0.107204         30.6   
405254                  0.009444      -0.20        0.439920         26.9   
405255                  0.025642      -0.20        0.275540         16.3   
405256                  0.181797       1.96        0.259254         16.5   
405257                  0.002647      -0.15       -0.033957         25.1   

        PE_ratio  ...  PB_ratio  NetProfitMargin_ratio  current_ratio  \
0          12.58  ...      3.19                   3.01           1.49   
1          11.39  ...      0.54                  15.70           0.92   
2          10.58  ...      2.60                  15.68           1.91   
3          11.09  ...      1.25                   3.17           1.60   
4           9.38  ...      0.81                  19.91           0.99   
...          ...  ...       ...                    ...            ...   
405253     23.98  ...      1.07                   1.29           1.27   
405254      4.40  ...      0.74                  11.58           0.92   
405255     12.02  ...      4.27                  12.06           1.74   
405256     16.74  ...      5.90                  21.12           1.24   
405257     19.34  ...      4.52                  17.41           1.08   

        roa_ratio  roe_ratio  difference_in_pr  ohe__sector_BANK  \
0            8.69      26.69         -2.067890               0.0   
1            0.67       5.54          6.037724               1.0   
2            3.39      25.78         -0.344570               1.0   
3            4.41      11.35         -2.231396               0.0   
4            0.81       8.91          0.178631               1.0   
...           ...        ...               ...               ...   
405253       1.18       4.87         -2.914038               0.0   
405254       2.16      17.79          3.950906               1.0   
405255       7.02      31.43         11.195938               0.0   
405256      13.83      37.37         10.915562               0.0   
405257       8.62      23.93         -1.199322               0.0   

        ohe__sector_FMCG  ohe__sector_RETAIL  ohe__sector_TECH  
0                    0.0                 1.0               0.0  
1                    0.0                 0.0               0.0  
2                    0.0                 0.0               0.0  
3                    0.0                 1.0               0.0  
4                    0.0                 0.0               0.0  
...                  ...              

**Defining Regressors and Response**

In [ ]:
y=df['nominal_return']
x=df.drop(['nominal_return'],axis=1)

**Creating Correlation Matrix for Multicorrelation Detection**

In [ ]:
R=x.corr()
R

horizon (days)    amount  Volatility_Buy  \
horizon (days)                  1.000000 -0.002030       -0.000486   
amount                         -0.002030  1.000000       -0.000120   
Volatility_Buy                 -0.000486 -0.000120        1.000000   
Volatility_sell                 0.084652  0.000330        0.852705   
Sharpe Ratio                   -0.000486 -0.000120        1.000000   
expected_return (yearly)        0.556673 -0.001110        0.147667   
inflation                       0.000637  0.001318       -0.028073   
ESG_ranking                     0.002692 -0.000677       -0.053567   
PE_ratio                        0.003669 -0.000910        0.015321   
EPS_ratio                      -0.001509  0.000399       -0.233564   
PS_ratio                       -0.000259 -0.000404        0.202416   
PB_ratio                       -0.000031 -0.000904        0.249334   
NetProfitMargin_ratio          -0.001947  0.000587       -0.249384   
current_ratio                  -0.000537  0.000611        0.132009   
roa_ratio                      -0.001523 -0.000404       -0.174418   
roe_ratio                      -0.000913  0.000634       -0.336224   
difference_in_pr                0.218510  0.000918        0.007688   
ohe__sector_BANK                0.000866  0.002118        0.060505   
ohe__sector_FMCG                0.000834  0.000067       -0.409339   
ohe__sector_RETAIL              0.000596 -0.001387        0.082203   
ohe__sector_TECH               -0.001188 -0.000761        0.073894   

                          Volatility_sell  Sharpe Ratio  \
horizon (days)                   0.084652     -0.000486   
amount                           0.000330     -0.000120   
Volatility_Buy                   0.852705      1.000000   
Volatility_sell                  1.000000      0.852705   
Sharpe Ratio                     0.852705      1.000000   
expected_return (yearly)         0.177267      0.147667   
inflation                        0.005472     -0.028073   
ESG_ranking                     -0.084554     -0.053567   
PE_ratio                         0.014359      0.015321   
EPS_ratio                       -0.219637     -0.233564   
PS_ratio                         0.157122      0.202416   
PB_ratio                         0.213715      0.249334   
NetProfitMargin_ratio           -0.258395     -0.249384   
current_ratio                    0.119144      0.132009   
roa_ratio                       -0.154856     -0.174418   
roe_ratio                       -0.319713     -0.336224   
difference_in_pr                -0.002481      0.007688   
ohe__sector_BANK                 0.044979      0.060505   
ohe__sector_FMCG                -0.407653     -0.409339   
ohe__sector_RETAIL               0.100612      0.082203   
ohe__sector_TECH                 0.075028      0.073894   

                          expected_return (yearly)  inflation  ESG_ranking  \
horizon (days)                            0.556673   0.000637     0.002692   
amount                                   -0.001110   0.001318    -0.000677   
Volatility_Buy                            0.147667  -0.028073    -0.053567   
Volatility_sell                           0.177267   0.005472    -0.084554   
Sharpe Ratio                              0.147667  -0.028073    -0.053567   
expected_return (yearly)                  1.000000  -0.017545    -0.056224   
inflation                                -0.017545   1.000000     0.002660   
ESG_ranking                              -0.056224   0.002660     1.000000   
PE_ratio                                  0.101735  -0.012094     0.118032   
EPS_ratio                                 0.011550   0.080124    -0.106311   
PS_ratio                                  0.281466   0.067225     0.175648   
PB_ratio                                  0.343305   0.021117     0.102919   
NetProfitMargin_ratio                     0.066708   0.050553    -0.141383   
current_ratio                             0.159831   0.019587     0.060230   
roa_ratio       

1) Cheaking Determinant of R

In [ ]:
R_mat=R.values
det_corr=np.linalg.det(R_mat)
print('Determinant of correlation matrix is : ',det_corr)

Determinant of correlation matrix is :  2.4632587248767072e-21


**As determinant of correlation matrix is very close to 0 it suspects: presence of Multicollinearity**

**Lets Check Condition Number**

In [ ]:
eigenvalues=np.linalg.eigvals(R_mat)
condition_number=np.max(eigenvalues)/np.min(eigenvalues)
print('Condition Number is : ',condition_number)

Condition Number is :  1.2772153313191858e+17


**To get more precise idea about multicollinearity we need to compute VIF(Varience Inflation Factor)**

In [ ]:
from sklearn.linear_model import LinearRegression

def VIF(data, feature):
    x_new = data.drop(columns=[feature])
    y_new = data[feature]
    model = LinearRegression()
    model.fit(x_new, y_new)
    y_pred = model.predict(x_new)
    sst = np.sum((y_new - np.mean(y_new))**2)
    sse = np.sum((y_new - y_pred)**2)
    r_squared = 1 - (sse / sst)
    if np.isclose(r_squared, 1):
        return np.inf

    return 1 / (1 - r_squared)


In [ ]:
vif_df = pd.DataFrame({'Feature': x.columns, 'VIF': x.columns.to_series().apply(lambda col: VIF(x, col))})
vif_df = vif_df.sort_values(by="VIF", ascending=False)

print(vif_df)


                                           Feature       VIF
Volatility_Buy                      Volatility_Buy       inf
Sharpe Ratio                          Sharpe Ratio       inf
PS_ratio                                  PS_ratio  8.055240
roa_ratio                                roa_ratio  6.807841
NetProfitMargin_ratio        NetProfitMargin_ratio  5.928728
roe_ratio                                roe_ratio  4.746900
ohe__sector_TECH                  ohe__sector_TECH  4.745453
Volatility_sell                    Volatility_sell  4.025421
current_ratio                        current_ratio  3.962024
PB_ratio                                  PB_ratio  2.956641
ohe__sector_RETAIL              ohe__sector_RETAIL  2.798219
ohe__sector_BANK                  ohe__sector_BANK  2.333956
ohe__sector_FMCG                  ohe__sector_FMCG  2.296633
ESG_ranking                            ESG_ranking  2.278517
expected_return (yearly)  expected_return (yearly)  2.129783
horizon (days)          

**Instead of dropping lets combine features with VIF Inf using PCA**

In [ ]:
from sklearn.decomposition import PCA
pca=PCA(n_components=1)
vif_inf=x[['Sharpe Ratio','Volatility_Buy']]
x['combined_sharpe_vol_buy']=pca.fit_transform(vif_inf)


In [ ]:
x_n=x.drop(['Sharpe Ratio','Volatility_Buy'],axis=1)
x_n

horizon (days)  amount  Volatility_sell  expected_return (yearly)  \
0                    2     100         0.385748                  0.001437   
1                  330   15000         0.236350                  0.170969   
2                    7    3000         0.235491                  0.002824   
3                    5   20000         0.429340                  0.000094   
4                  360   15000         0.254011                  0.149979   
...                ...     ...              ...                       ...   
405253             480   50000         0.253235                  0.053763   
405254             300   50000         0.535448                  0.009444   
405255             210    3000         0.359625                  0.025642   
405256             270     500         0.218904                  0.181797   
405257              15    2000         0.145768                  0.002647   

        inflation  ESG_ranking  PE_ratio  EPS_ratio  PS_ratio  PB_ratio  \
0            1.96         12.0     12.58       3.73      0.38      3.19   
1           -0.20         26.3     11.39       1.26      1.71      0.54   
2           -0.20         19.8     10.58       5.64      1.67      2.60   
3           -0.20         12.9     11.09       3.27      0.36      1.25   
4           -0.50         27.9      9.38       5.46      1.87      0.81   
...           ...          ...       ...        ...       ...       ...   
405253      -0.15         30.6     23.98       1.19      0.31      1.07   
405254      -0.20         26.9      4.40       2.16      0.50      0.74   
405255      -0.20         16.3     12.02       3.71      1.46      4.27   
405256       1.96         16.5     16.74       2.43      3.55      5.90   
405257      -0.15         25.1     19.34       1.80      3.37      4.52   

        NetProfitMargin_ratio  current_ratio  roa_ratio  roe_ratio  \
0                        3.01           1.49       8.69      26.69   
1                       15.70           0.92       0.67       5.54   
2                       15.68           1.91       3.39      25.78   
3                        3.17           1.60       4.41      11.35   
4                       19.91           0.99       0.81       8.91   
...                       ...            ...        ...        ...   
405253                   1.29           1.27       1.18       4.87   
405254                  11.58           0.92       2.16      17.79   
405255                  12.06           1.74       7.02      31.43   
405256                  21.12           1.24      13.83      37.37   
405257                  17.41           1.08       8.62      23.93   

        difference_in_pr  ohe__sector_BANK  ohe__sector_FMCG  \
0              -2.067890               0.0               0.0   
1               6.037724               1.0               0.0   
2              -0.344570               1.0               0.0   
3              -2.231396               0.0               0.0   
4               0.178631               1.0               0.0   
...                  ...               ...               ...   
405253         -2.914038               0.0               0.0   
405254          3.950906               1.0               0.0   
405255         11.195938               0.0               0.0   
405256         10.915562               0.0               0.0   
405257         -1.199322               0.0               1.0   

        ohe__sector_RETAIL  ohe__sector_TECH  combined_sharpe_vol_buy  
0                      1.0               0.0                 0.183179  
1                      0.0               0.0                 0.097115  
2                      0.0               0.0                -0.021915  
3                      1.0               0.0                 0.246669  
4                      0.0               0.0                -0.084183  
...                    ...               ...                      ...  
405253                 0.0               0.0                -0.01843

**Now lets check Multicollinearity after combining the features**

In [ ]:
vif_df_n= pd.DataFrame({'Feature': x_n.columns, 'VIF': x_n.columns.to_series().apply(lambda col: VIF(x_n, col))})
vif_df_n = vif_df_n.sort_values(by="VIF", ascending=False)

print(vif_df_n)


                                           Feature       VIF
PS_ratio                                  PS_ratio  8.055240
roa_ratio                                roa_ratio  6.807841
NetProfitMargin_ratio        NetProfitMargin_ratio  5.928728
roe_ratio                                roe_ratio  4.746900
ohe__sector_TECH                  ohe__sector_TECH  4.745453
combined_sharpe_vol_buy    combined_sharpe_vol_buy  4.187674
Volatility_sell                    Volatility_sell  4.025421
current_ratio                        current_ratio  3.962024
PB_ratio                                  PB_ratio  2.956641
ohe__sector_RETAIL              ohe__sector_RETAIL  2.798219
ohe__sector_BANK                  ohe__sector_BANK  2.333956
ohe__sector_FMCG                  ohe__sector_FMCG  2.296633
ESG_ranking                            ESG_ranking  2.278517
expected_return (yearly)  expected_return (yearly)  2.129783
horizon (days)                      horizon (days)  1.674176
PE_ratio                

In [ ]:
vif_high=x[['PS_ratio','roa_ratio','NetProfitMargin_ratio']]
x_n['combined_ps_roa_net']=pca.fit_transform(vif_high)
x_n_2=x_n.drop(['PS_ratio','roa_ratio','NetProfitMargin_ratio'],axis=1)

vif_df_n_2= pd.DataFrame({'Feature': x_n_2.columns, 'VIF': x_n_2.columns.to_series().apply(lambda col: VIF(x_n_2, col))})
vif_df_n_2 = vif_df_n_2.sort_values(by="VIF", ascending=False)

print(vif_df_n_2)



                                           Feature       VIF
ohe__sector_TECH                  ohe__sector_TECH  4.136793
combined_sharpe_vol_buy    combined_sharpe_vol_buy  4.071900
combined_ps_roa_net            combined_ps_roa_net  4.024756
Volatility_sell                    Volatility_sell  3.985962
roe_ratio                                roe_ratio  2.796387
ohe__sector_RETAIL              ohe__sector_RETAIL  2.462716
ESG_ranking                            ESG_ranking  2.185324
expected_return (yearly)  expected_return (yearly)  2.126480
ohe__sector_FMCG                  ohe__sector_FMCG  2.096807
ohe__sector_BANK                  ohe__sector_BANK  2.035497
current_ratio                        current_ratio  1.890181
PB_ratio                                  PB_ratio  1.880508
horizon (days)                      horizon (days)  1.674031
PE_ratio                                  PE_ratio  1.364706
difference_in_pr                  difference_in_pr  1.352716
EPS_ratio               

In [ ]:
x_n_2.to_csv('/content/drive/MyDrive/Neelu project/final_df.csv')
x_n_2

horizon (days)  amount  Volatility_sell  expected_return (yearly)  \
0                    2     100         0.385748                  0.001437   
1                  330   15000         0.236350                  0.170969   
2                    7    3000         0.235491                  0.002824   
3                    5   20000         0.429340                  0.000094   
4                  360   15000         0.254011                  0.149979   
...                ...     ...              ...                       ...   
405253             480   50000         0.253235                  0.053763   
405254             300   50000         0.535448                  0.009444   
405255             210    3000         0.359625                  0.025642   
405256             270     500         0.218904                  0.181797   
405257              15    2000         0.145768                  0.002647   

        inflation  ESG_ranking  PE_ratio  EPS_ratio  PB_ratio  current_ratio  \
0            1.96         12.0     12.58       3.73      3.19           1.49   
1           -0.20         26.3     11.39       1.26      0.54           0.92   
2           -0.20         19.8     10.58       5.64      2.60           1.91   
3           -0.20         12.9     11.09       3.27      1.25           1.60   
4           -0.50         27.9      9.38       5.46      0.81           0.99   
...           ...          ...       ...        ...       ...            ...   
405253      -0.15         30.6     23.98       1.19      1.07           1.27   
405254      -0.20         26.9      4.40       2.16      0.74           0.92   
405255      -0.20         16.3     12.02       3.71      4.27           1.74   
405256       1.96         16.5     16.74       2.43      5.90           1.24   
405257      -0.15         25.1     19.34       1.80      4.52           1.08   

        roe_ratio  difference_in_pr  ohe__sector_BANK  ohe__sector_FMCG  \
0           26.69         -2.067890               0.0               0.0   
1            5.54          6.037724               1.0               0.0   
2           25.78         -0.344570               1.0               0.0   
3           11.35         -2.231396               0.0               0.0   
4            8.91          0.178631               1.0               0.0   
...           ...               ...               ...               ...   
405253       4.87         -2.914038               0.0               0.0   
405254      17.79          3.950906               1.0               0.0   
405255      31.43         11.195938               0.0               0.0   
405256      37.37         10.915562               0.0               0.0   
405257      23.93         -1.199322               0.0               1.0   

        ohe__sector_RETAIL  ohe__sector_TECH  combined_sharpe_vol_buy  \
0                      1.0               0.0                 0.183179   
1                      0.0               0.0                 0.097115   
2                      0.0               0.0                -0.021915   
3                      1.0               0.0                 0.246669   
4                      0.0               0.0                -0.084183   
...                    ...               ...                      ...   
405253                 0.0               0.0                -0.018435   
405254                 0.0               0.0                 0.221087   
405255                 0.0               0.0                 0.087797   
405256                 0.0               1.0                -0.113768   
405257                 0.0               0.0                -0.173548   

        combined_ps_roa_net  
0                 -4.611268  
1                  3.595783  
2                  4.710001  
3                 -6.262251  
4                  7.447274  
...                     ...  
405253            -9.304397  
405254             0.345296  
405255             2.956594  
405256            14.237683  
405257        

In [ ]:
df_nn=pd.concat([x_n_2,y],axis=1)
df_nn.to_csv('/content/drive/MyDrive/Neelu project/final_df_with_y.csv')


In [ ]:
model_1=LinearRegression()
model_1.fit(x_n_2,y)
y_pred_1 = model_1.predict(x_n_2)
sst_1 = np.sum((y - np.mean(y))**2)
sse_1 = np.sum((y - y_pred_1)**2)
r_squared_1 = 1 - (sse_1 / sst_1)
print('R-squared value is : ',r_squared_1)


R-squared value is :  0.41119832923600097


In [ ]:
from sklearn.linear_model import Ridge

ridge_model = Ridge(alpha=1.0)
ridge_model.fit(x_n_2, y)

print("Ridge Regression Coefficients:", ridge_model.coef_)
y_pred_2 = ridge_model.predict(x_n_2)
sst_2 = np.sum((y - np.mean(y))**2)
sse_2 = np.sum((y - y_pred_2)**2)
r_squared_2= 1 - (sse_2 / sst_2)
print('R-squared value is : ',r_squared_2)


Ridge Regression Coefficients: [-5.26472614e-05  1.14973459e-08 -7.31569950e-01  1.09534944e+00
  1.71469446e-04 -2.40217705e-03 -1.09945374e-04 -3.90110932e-03
 -1.10492866e-02 -3.89463777e-03  1.58443209e-04  1.26678336e-03
  1.20327623e-02  9.30411453e-02  1.20245377e-02  1.11799434e-01
  6.18076033e-01 -1.45051543e-03]
R-squared value is :  0.41119828679828085


In [ ]:
import numpy as np
import statsmodels.api as sm

# Compute mean and standard deviation for each feature in X
mean_x = x_n_2.mean()
std_x = x_n_2.std()

# Define 3-sigma limits for each feature
lower_limit = mean_x - (1.7)* std_x
upper_limit = mean_x + (1.7) * std_x

# Create mask: Keep only rows where all features are within 3-sigma limits
mask = ((x_n_2 >= lower_limit) & (x_n_2 <= upper_limit)).all(axis=1)
# Apply the mask to X and y
x_n_2_filtered = x_n_2[mask]
y_filtered = y[mask]

# Add constant term for intercept
X_wls = sm.add_constant(x_n_2_filtered)

# Compute weights (avoid division by zero)
weights = 1 / (np.abs(y_filtered) + 1e-5)

# Fit Weighted Least Squares (WLS)
wls_model = sm.WLS(y_filtered, X_wls, weights=weights).fit()

# Print summary
print(wls_model.summary())


                            WLS Regression Results                            
Dep. Variable:         nominal_return   R-squared:                       0.863
Model:                            WLS   Adj. R-squared:                  0.863
Method:                 Least Squares   F-statistic:                 2.306e+04
Date:                Wed, 16 Apr 2025   Prob (F-statistic):               0.00
Time:                        07:56:31   Log-Likelihood:             1.3007e+05
No. Observations:               51285   AIC:                        -2.601e+05
Df Residuals:                   51270   BIC:                        -2.600e+05
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
const                   

In [ ]:
# Compute mean and standard deviation for each feature in X
mean_x = x_n_2.mean()
std_x = x_n_2.std()

# Define 3-sigma limits for each feature
lower_limit = mean_x - (3)* std_x
upper_limit = mean_x + (3) * std_x

# Create mask: Keep only rows where all features are within 3-sigma limits
mask = ((x_n_2 >= lower_limit) & (x_n_2 <= upper_limit)).all(axis=1)
# Apply the mask to X and y
x_n_2_filtered = x_n_2[mask]
y_filtered = y[mask]

# Add constant term for intercept
X_ols = sm.add_constant(x_n_2_filtered)

model_3=LinearRegression()
model_3.fit(x_n_2_filtered,y_filtered)
y_pred_3 = model_3.predict(x_n_2_filtered)
sst_3 = np.sum((y_filtered - np.mean(y_filtered))**2)
sse_3 = np.sum((y_filtered - y_pred_3)**2)
r_squared_3 = 1 - (sse_3 / sst_3)
print('R-squared value is : ',r_squared_3)



R-squared value is :  0.5730507154344191


**The above analysis shows that the main concern for poor model performance is outliers.**
**So we'll now handle outliers using various techniques and look for improvement in the model**

### **Handling Outliers Using the Hat Matrix in Regression**



In [ ]:
import numpy as np
import statsmodels.api as sm

X = np.asarray(sm.add_constant(x_n_2))
XtX_inv = np.linalg.pinv(X.T @ X)
H_diag = np.einsum('ij,ji->i', X @ XtX_inv, X.T)  # Efficient diagonal extraction

print(H_diag)


[2.51780722e-05 2.67087083e-05 2.47617442e-05 ... 3.33723733e-05
 3.27664981e-05 2.83862511e-05]


In [ ]:
p=X.shape[1]-1
n=X.shape[0]
threshold=(2*(p+1))/n
high_leverage_points=np.where(H_diag>threshold)[0]
print(high_leverage_points)
len(high_leverage_points)

[     6     18     31 ... 405215 405221 405249]


33753

### Now we have high levarage points but not all are influencial so we'll first use cooks distance to check for influencial points ###

In [ ]:
from statsmodels.stats.outliers_influence import OLSInfluence

model = sm.OLS(y, X).fit()
influence = OLSInfluence(model)
cooks_d = influence.cooks_distance[0]



cooks_threshold = 4 / n

influential_points = np.where(cooks_d > cooks_threshold)[0]


print("Influential points (indices):", influential_points)


Influential points (indices): [    18     39     45 ... 405240 405249 405254]


In [ ]:
common_points = np.intersect1d(high_leverage_points, influential_points)
print("Points that are both high leverage and influential:", common_points)
len(common_points)

Points that are both high leverage and influential: [    18     39    163 ... 405187 405190 405249]


11808

### So there are 11808 high leverage and influential points we'll remove them try to fit the new model ###

In [ ]:

import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

X_cleaned = np.delete(X, influential_points, axis=0)
y_cleaned = np.delete(y, influential_points, axis=0)

feature_names = ["Intercept"] + list(x_n_2.columns)
X_cleaned_df = pd.DataFrame(X_cleaned, columns=feature_names)

y_cleaned = y_cleaned.ravel()

model_cleaned = LinearRegression()
model_cleaned.fit(X_cleaned_df, y_cleaned)

y_pred_cleaned = model_cleaned.predict(X_cleaned_df)
r_squared_cleaned = r2_score(y_cleaned, y_pred_cleaned)

print("R-squared value is:", r_squared_cleaned)




R-squared value is: 0.5553172980254131


In [ ]:
x_n_2

horizon (days)  amount  Volatility_sell  expected_return (yearly)  \
0                    2     100         0.385748                  0.001437   
1                  330   15000         0.236350                  0.170969   
2                    7    3000         0.235491                  0.002824   
3                    5   20000         0.429340                  0.000094   
4                  360   15000         0.254011                  0.149979   
...                ...     ...              ...                       ...   
405253             480   50000         0.253235                  0.053763   
405254             300   50000         0.535448                  0.009444   
405255             210    3000         0.359625                  0.025642   
405256             270     500         0.218904                  0.181797   
405257              15    2000         0.145768                  0.002647   

        inflation  ESG_ranking  PE_ratio  EPS_ratio  PB_ratio  current_ratio  \
0            1.96         12.0     12.58       3.73      3.19           1.49   
1           -0.20         26.3     11.39       1.26      0.54           0.92   
2           -0.20         19.8     10.58       5.64      2.60           1.91   
3           -0.20         12.9     11.09       3.27      1.25           1.60   
4           -0.50         27.9      9.38       5.46      0.81           0.99   
...           ...          ...       ...        ...       ...            ...   
405253      -0.15         30.6     23.98       1.19      1.07           1.27   
405254      -0.20         26.9      4.40       2.16      0.74           0.92   
405255      -0.20         16.3     12.02       3.71      4.27           1.74   
405256       1.96         16.5     16.74       2.43      5.90           1.24   
405257      -0.15         25.1     19.34       1.80      4.52           1.08   

        roe_ratio  difference_in_pr  ohe__sector_BANK  ohe__sector_FMCG  \
0           26.69         -2.067890               0.0               0.0   
1            5.54          6.037724               1.0               0.0   
2           25.78         -0.344570               1.0               0.0   
3           11.35         -2.231396               0.0               0.0   
4            8.91          0.178631               1.0               0.0   
...           ...               ...               ...               ...   
405253       4.87         -2.914038               0.0               0.0   
405254      17.79          3.950906               1.0               0.0   
405255      31.43         11.195938               0.0               0.0   
405256      37.37         10.915562               0.0               0.0   
405257      23.93         -1.199322               0.0               1.0   

        ohe__sector_RETAIL  ohe__sector_TECH  combined_sharpe_vol_buy  \
0                      1.0               0.0                 0.183179   
1                      0.0               0.0                 0.097115   
2                      0.0               0.0                -0.021915   
3                      1.0               0.0                 0.246669   
4                      0.0               0.0                -0.084183   
...                    ...               ...                      ...   
405253                 0.0               0.0                -0.018435   
405254                 0.0               0.0                 0.221087   
405255                 0.0               0.0                 0.087797   
405256                 0.0               1.0                -0.113768   
405257                 0.0               0.0                -0.173548   

        combined_ps_roa_net  
0                 -4.611268  
1                  3.595783  
2                  4.710001  
3                 -6.262251  
4                  7.447274  
...                     ...  
405253            -9.304397  
405254             0.345296  
405255             2.956594  
405256            14.237683  
405257        

In [ ]:
from numpy.linalg import pinv

from scipy.stats import chi2
from scipy.spatial.distance import mahalanobis

X_df = pd.DataFrame(X_cleaned)

# Compute Mahalanobis distance with pseudo-inverse or  Moore–Penrose inverse
cov_matrix = np.cov(X_df.values, rowvar=False)
inv_cov_matrix = np.linalg.pinv(cov_matrix)
mean_distr = X_df.mean(axis=0).values

mahalanobis_distances = X_df.apply(lambda row: mahalanobis(row.values, mean_distr, inv_cov_matrix), axis=1)

threshold = chi2.ppf(0.85, df=X_df.shape[1])
inliers = mahalanobis_distances < np.sqrt(threshold)

X_filtered = X_df[inliers]
y_filtered = y_cleaned[inliers]
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

# Create and fit the model
model = LinearRegression()
model.fit(X_filtered, y_filtered)

# Predict on the filtered data
y_pred = model.predict(X_filtered)

# Calculate R² score
r2 = r2_score(y_filtered, y_pred)
print(f"R² Score on filtered data: {r2:.4f}")

R² Score on filtered data: 0.6654


In [ ]:
from sklearn.ensemble import IsolationForest
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
import numpy as np
import pandas as pd

# Ensure consistent types
if isinstance(X_filtered, pd.DataFrame):
    y_filtered = pd.Series(y_filtered) if not isinstance(y_filtered, pd.Series) else y_filtered

# Fit Isolation Forest
iso_forest = IsolationForest(contamination=0.05, random_state=42)
outlier_preds = iso_forest.fit_predict(X_filtered)

# Filter and reset indices
if isinstance(X_filtered, pd.DataFrame):
    X_clean = X_filtered[outlier_preds == 1].reset_index(drop=True)
    y_clean = y_filtered[outlier_preds == 1].reset_index(drop=True)
else:
    X_clean = X_filtered[outlier_preds == 1]
    y_clean = y_filtered[outlier_preds == 1]

# Cross-validation
model_i = LinearRegression()
scores = cross_val_score(model_i, X_clean, y_clean, cv=7)
print("Cross-val scores:", scores)
print("Mean score:", scores.mean())

Cross-val scores: [0.74765727 0.74986786 0.7516848  0.75521169 0.74923555 0.7554519
 0.75322124]
Mean score: 0.7517614713579098


In [ ]:
# Rename the first column of df2 to "intercept"
X_clean = X_clean.rename(columns={X_clean.columns[0]: "intercept"})

# Rename the remaining 18 columns to match df1's columns
X_clean.columns = ["intercept"] + list(x_n_2.columns)
X_clean

intercept  horizon (days)   amount  Volatility_sell  \
0             1.0             2.0    100.0         0.385748   
1             1.0           330.0  15000.0         0.236350   
2             1.0             7.0   3000.0         0.235491   
3             1.0             5.0  20000.0         0.429340   
4             1.0           360.0  15000.0         0.254011   
...           ...             ...      ...              ...   
309884        1.0            90.0   5000.0         0.186519   
309885        1.0           480.0  50000.0         0.253235   
309886        1.0           210.0   3000.0         0.359625   
309887        1.0           270.0    500.0         0.218904   
309888        1.0            15.0   2000.0         0.145768   

        expected_return (yearly)  inflation  ESG_ranking  PE_ratio  EPS_ratio  \
0                       0.001437       1.96         12.0     12.58       3.73   
1                       0.170969      -0.20         26.3     11.39       1.26   
2                       0.002824      -0.20         19.8     10.58       5.64   
3                       0.000094      -0.20         12.9     11.09       3.27   
4                       0.149979      -0.50         27.9      9.38       5.46   
...                          ...        ...          ...       ...        ...   
309884                 -0.009739       1.68         24.8     36.50       0.40   
309885                  0.053763      -0.15         30.6     23.98       1.19   
309886                  0.025642      -0.20         16.3     12.02       3.71   
309887                  0.181797       1.96         16.5     16.74       2.43   
309888                  0.002647      -0.15         25.1     19.34       1.80   

        PB_ratio  current_ratio  roe_ratio  difference_in_pr  \
0           3.19           1.49      26.69         -2.067890   
1           0.54           0.92       5.54          6.037724   
2           2.60           1.91      25.78         -0.344570   
3           1.25           1.60      11.35         -2.231396   
4           0.81           0.99       8.91          0.178631   
...          ...            ...        ...               ...   
309884      1.06           0.95       0.39          0.311943   
309885      1.07           1.27       4.87         -2.914038   
309886      4.27           1.74      31.43         11.195938   
309887      5.90           1.24      37.37         10.915562   
309888      4.52           1.08      23.93         -1.199322   

        ohe__sector_BANK  ohe__sector_FMCG  ohe__sector_RETAIL  \
0                    0.0               0.0                 1.0   
1                    1.0               0.0                 0.0   
2                    1.0               0.0                 0.0   
3                    0.0               0.0                 1.0   
4                    1.0               0.0                 0.0   
...                  ...               ...                 ...   
309884               1.0               0.0                 0.0   
309885               0.0               0.0                 0.0   
309886               0.0               0.0                 0.0   
309887               0.0               0.0                 0.0   
309888               0.0               1.0                 0.0   

        ohe__sector_TECH  combined_sharpe_vol_buy  combined_ps_roa_net  
0                    0.0                 0.183179            -4.611268  
1                    0.0                 0.097115             3.595783  
2                    0.0                -0.021915             4.710001  
3                    0.0                 0.246669            -6.262251  
4                    0.0                -0.084183             7.447274  
...                  ...                      ...                  ...  
309884               0.0                -0.101316           -10.205722  
309885               0.0                -0.018435            -9.304397  
309886               0.0                 0.087797          

In [ ]:
y_clean

0        -0.037225
1         0.324317
2        -0.005756
3        -0.058388
4         0.003444
            ...   
309884    0.022654
309885   -0.107204
309886    0.275540
309887    0.259254
309888   -0.033957
Length: 309889, dtype: float64

In [ ]:
final_df=pd.concat([X_clean,y_clean],axis=1)
# Rename the last column (position 19 if 0-indexed)
final_df.rename(columns={final_df.columns[-1]: 'Nominal_return'}, inplace=True)
final_df

intercept  horizon (days)   amount  Volatility_sell  \
0             1.0             2.0    100.0         0.385748   
1             1.0           330.0  15000.0         0.236350   
2             1.0             7.0   3000.0         0.235491   
3             1.0             5.0  20000.0         0.429340   
4             1.0           360.0  15000.0         0.254011   
...           ...             ...      ...              ...   
309884        1.0            90.0   5000.0         0.186519   
309885        1.0           480.0  50000.0         0.253235   
309886        1.0           210.0   3000.0         0.359625   
309887        1.0           270.0    500.0         0.218904   
309888        1.0            15.0   2000.0         0.145768   

        expected_return (yearly)  inflation  ESG_ranking  PE_ratio  EPS_ratio  \
0                       0.001437       1.96         12.0     12.58       3.73   
1                       0.170969      -0.20         26.3     11.39       1.26   
2                       0.002824      -0.20         19.8     10.58       5.64   
3                       0.000094      -0.20         12.9     11.09       3.27   
4                       0.149979      -0.50         27.9      9.38       5.46   
...                          ...        ...          ...       ...        ...   
309884                 -0.009739       1.68         24.8     36.50       0.40   
309885                  0.053763      -0.15         30.6     23.98       1.19   
309886                  0.025642      -0.20         16.3     12.02       3.71   
309887                  0.181797       1.96         16.5     16.74       2.43   
309888                  0.002647      -0.15         25.1     19.34       1.80   

        PB_ratio  current_ratio  roe_ratio  difference_in_pr  \
0           3.19           1.49      26.69         -2.067890   
1           0.54           0.92       5.54          6.037724   
2           2.60           1.91      25.78         -0.344570   
3           1.25           1.60      11.35         -2.231396   
4           0.81           0.99       8.91          0.178631   
...          ...            ...        ...               ...   
309884      1.06           0.95       0.39          0.311943   
309885      1.07           1.27       4.87         -2.914038   
309886      4.27           1.74      31.43         11.195938   
309887      5.90           1.24      37.37         10.915562   
309888      4.52           1.08      23.93         -1.199322   

        ohe__sector_BANK  ohe__sector_FMCG  ohe__sector_RETAIL  \
0                    0.0               0.0                 1.0   
1                    1.0               0.0                 0.0   
2                    1.0               0.0                 0.0   
3                    0.0               0.0                 1.0   
4                    1.0               0.0                 0.0   
...                  ...               ...                 ...   
309884               1.0               0.0                 0.0   
309885               0.0               0.0                 0.0   
309886               0.0               0.0                 0.0   
309887               0.0               0.0                 0.0   
309888               0.0               1.0                 0.0   

        ohe__sector_TECH  combined_sharpe_vol_buy  combined_ps_roa_net  \
0                    0.0                 0.183179            -4.611268   
1                    0.0                 0.097115             3.595783   
2                    0.0                -0.021915             4.710001   
3                    0.0                 0.246669            -6.262251   
4                    0.0                -0.084183             7.447274   
...                  ...                      ...                  ...   
309884               0.0                -0.101316           -10.205722   
309885               0.0                -0.018435            -9.304397   
309886               0.0                 0.087797 

In [ ]:
X_check=final_df.drop(['Nominal_return'],axis=1)
y_check=final_df['Nominal_return']
model_check=LinearRegression()
score_check=cross_val_score(model_check,X_check,y_check,cv=7)
print("Cross-val scores:", score_check)
print("Mean score:", score_check.mean())
model_check.fit(X_check, y_check)

# Access intercept and coefficients
intercept = model_check.intercept_
coefficients = model_check.coef_

# Print results
print("\nIntercept:", intercept)
print("Coefficients:")
for feature, coef in zip(X_check.columns, coefficients):
    print(f"{feature}: {coef}")

y_pred = model_check.predict(X_check)
residuals = y_check - y_pred

# RSS
RSS = np.sum(residuals ** 2)

# n = number of observations
n = len(y_check)

# p = number of predictors + 1 (intercept)
p = X_check.shape[1] + 1

# Estimate of sigma^2 from full model
sigma_squared = RSS / (n - p)

# Now compute Mallows' Cp for this model (full model)
Cp = RSS / sigma_squared - n + 2 * p

print("Mallows' Cp for the model:", Cp)

Cross-val scores: [0.74765727 0.74986786 0.7516848  0.75521169 0.74923555 0.7554519
 0.75322124]
Mean score: 0.7517614713579098

Intercept: 0.21977243177312514
Coefficients:
intercept: 0.0
horizon (days): -5.914093145341658e-05
amount: 4.81050430864436e-09
Volatility_sell: -0.5957034035928511
expected_return (yearly): 0.4296089325548284
inflation: -0.001291718597708309
ESG_ranking: -0.0008977311850272848
PE_ratio: -0.00043650095397450344
EPS_ratio: -0.005111170751859304
PB_ratio: -0.0052147611439107055
current_ratio: -0.0016198759780238691
roe_ratio: 2.1373858283935063e-05
difference_in_pr: 0.014184179871227125
ohe__sector_BANK: -0.015432751609644411
ohe__sector_FMCG: 0.015503802193836088
ohe__sector_RETAIL: -0.005235405131084042
ohe__sector_TECH: -0.012978572523972805
combined_sharpe_vol_buy: 0.49449107680936977
combined_ps_roa_net: 0.0017874429912167645
Mallows' Cp for the model: 20.0


In [ ]:
final_df.to_csv('/content/drive/MyDrive/Neelu project/saaf_data.csv')

In [ ]:
final_df1=final_df.drop(['intercept'],axis=1)
final_df1

horizon (days)   amount  Volatility_sell  expected_return (yearly)  \
0                  2.0    100.0         0.385748                  0.001437   
1                330.0  15000.0         0.236350                  0.170969   
2                  7.0   3000.0         0.235491                  0.002824   
3                  5.0  20000.0         0.429340                  0.000094   
4                360.0  15000.0         0.254011                  0.149979   
...                ...      ...              ...                       ...   
309884            90.0   5000.0         0.186519                 -0.009739   
309885           480.0  50000.0         0.253235                  0.053763   
309886           210.0   3000.0         0.359625                  0.025642   
309887           270.0    500.0         0.218904                  0.181797   
309888            15.0   2000.0         0.145768                  0.002647   

        inflation  ESG_ranking  PE_ratio  EPS_ratio  PB_ratio  current_ratio  \
0            1.96         12.0     12.58       3.73      3.19           1.49   
1           -0.20         26.3     11.39       1.26      0.54           0.92   
2           -0.20         19.8     10.58       5.64      2.60           1.91   
3           -0.20         12.9     11.09       3.27      1.25           1.60   
4           -0.50         27.9      9.38       5.46      0.81           0.99   
...           ...          ...       ...        ...       ...            ...   
309884       1.68         24.8     36.50       0.40      1.06           0.95   
309885      -0.15         30.6     23.98       1.19      1.07           1.27   
309886      -0.20         16.3     12.02       3.71      4.27           1.74   
309887       1.96         16.5     16.74       2.43      5.90           1.24   
309888      -0.15         25.1     19.34       1.80      4.52           1.08   

        roe_ratio  difference_in_pr  ohe__sector_BANK  ohe__sector_FMCG  \
0           26.69         -2.067890               0.0               0.0   
1            5.54          6.037724               1.0               0.0   
2           25.78         -0.344570               1.0               0.0   
3           11.35         -2.231396               0.0               0.0   
4            8.91          0.178631               1.0               0.0   
...           ...               ...               ...               ...   
309884       0.39          0.311943               1.0               0.0   
309885       4.87         -2.914038               0.0               0.0   
309886      31.43         11.195938               0.0               0.0   
309887      37.37         10.915562               0.0               0.0   
309888      23.93         -1.199322               0.0               1.0   

        ohe__sector_RETAIL  ohe__sector_TECH  combined_sharpe_vol_buy  \
0                      1.0               0.0                 0.183179   
1                      0.0               0.0                 0.097115   
2                      0.0               0.0                -0.021915   
3                      1.0               0.0                 0.246669   
4                      0.0               0.0                -0.084183   
...                    ...               ...                      ...   
309884                 0.0               0.0                -0.101316   
309885                 0.0               0.0                -0.018435   
309886                 0.0               0.0                 0.087797   
309887                 0.0               1.0                -0.113768   
309888                 0.0               0.0                -0.173548   

        combined_ps_roa_net  Nominal_return  
0                 -4.611268       -0.037225  
1                  3.595783        0.324317  
2                  4.710001       -0.005756  
3                 -6.262251       -0.058388  
4                  7.447274        0.003444  
...                     ...             ...  
309884    

In [ ]:
import statsmodels.api as sm

def forward_selection(X, y):
    remaining = list(X.columns)
    selected = []
    current_score, best_new_score = -float('inf'), -float('inf')

    while remaining:
        scores_with_candidates = []
        for candidate in remaining:
            features = selected + [candidate]
            X_with_const = sm.add_constant(X[features], has_constant='add')
            model = sm.OLS(y, X_with_const).fit()
            score = model.rsquared_adj
            scores_with_candidates.append((score, candidate))

        scores_with_candidates.sort(reverse=True)
        best_new_score, best_candidate = scores_with_candidates[0]

        if best_new_score > current_score:
            remaining.remove(best_candidate)
            selected.append(best_candidate)
            current_score = best_new_score
        else:
            break

    X_final = sm.add_constant(X[selected], has_constant='add')
    final_model = sm.OLS(y, X_final).fit()
    return final_model, selected


In [ ]:
model, selected_features = forward_selection(X_check, y_check)
print(model.summary())
print("Selected features:", selected_features)

# Get residual sum of squares from forward-selected model
RSS_fs = np.sum(model.resid ** 2)

# Now fit the full model to get sigma^2
X_full_const = sm.add_constant(X_check, has_constant='add')
full_model = sm.OLS(y_check, X_full_const).fit()
RSS_full = np.sum(full_model.resid ** 2)

# Parameters
n = len(y_check)
p_fs = len(selected_features) + 1  # +1 for intercept
p_full = X_check.shape[1] + 1

# Estimate sigma^2 from full model
sigma_squared = RSS_full / (n - p_full)

# Mallows' Cp for the forward-selected model
Cp_fs = RSS_fs / sigma_squared - n + 2 * p_fs

print("Selected Features:", selected_features)
print("Mallows' Cp for Forward Selected Model:", Cp_fs)


                            OLS Regression Results                            
Dep. Variable:         Nominal_return   R-squared:                       0.752
Model:                            OLS   Adj. R-squared:                  0.752
Method:                 Least Squares   F-statistic:                 5.521e+04
Date:                Wed, 16 Apr 2025   Prob (F-statistic):               0.00
Time:                        07:58:02   Log-Likelihood:             3.5834e+05
No. Observations:              309889   AIC:                        -7.166e+05
Df Residuals:                  309871   BIC:                        -7.164e+05
Df Model:                          17                                         
Covariance Type:            nonrobust                                         
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
const                   

In [ ]:
x_2=final_df1.drop(['Nominal_return'],axis=1)
y_2=final_df1['Nominal_return']

In [ ]:
import statsmodels.api as sm

def backward_elimination(X, y, significance_level=0.05):
    features = list(X.columns)

    while True:
        X_with_const = sm.add_constant(X[features], has_constant='add')
        model = sm.OLS(y, X_with_const).fit()
        pvalues = model.pvalues.drop('const', errors='ignore')  # exclude intercept

        max_pval = pvalues.max()
        if max_pval > significance_level:
            worst_feature = pvalues.idxmax()
            features.remove(worst_feature)
        else:
            break

    final_model = sm.OLS(y, sm.add_constant(X[features], has_constant='add')).fit()
    return final_model, features


In [ ]:
model, selected_features = backward_elimination(x_2, y_2)
print(model.summary())
print("Selected features:", selected_features)
# Get residual sum of squares from forward-selected model
RSS_fs = np.sum(model.resid ** 2)

# Now fit the full model to get sigma^2
X_full_const = sm.add_constant(x_2, has_constant='add')
full_model = sm.OLS(y_check, X_full_const).fit()
RSS_full = np.sum(full_model.resid ** 2)

# Parameters
n = len(y_2)
p_fs = len(selected_features) + 1  # +1 for intercept
p_full = X_check.shape[1] + 1

# Estimate sigma^2 from full model
sigma_squared = RSS_full / (n - p_full)

# Mallows' Cp for the forward-selected model
Cp_fs = RSS_fs / sigma_squared - n + 2 * p_fs

print("Selected Features:", selected_features)
print("Mallows' Cp for Forward Selected Model:", Cp_fs)





                            OLS Regression Results                            
Dep. Variable:         Nominal_return   R-squared:                       0.752
Model:                            OLS   Adj. R-squared:                  0.752
Method:                 Least Squares   F-statistic:                 5.867e+04
Date:                Wed, 16 Apr 2025   Prob (F-statistic):               0.00
Time:                        07:58:08   Log-Likelihood:             3.5834e+05
No. Observations:              309889   AIC:                        -7.166e+05
Df Residuals:                  309872   BIC:                        -7.165e+05
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
const                   

In [ ]:
from sklearn.linear_model import Ridge, Lasso
from sklearn.model_selection import cross_val_score
from sklearn.metrics import r2_score
import numpy as np

# Ridge Regression
ridge = Ridge(alpha=0.001)  # You can tune alpha with cross-validation
ridge.fit(X_check, y_check)
y_pred_ridge = ridge.predict(X_check)
ridge_r2 = r2_score(y_check, y_pred_ridge)
print("Ridge Regression R² Score:", ridge_r2)

# Lasso Regression
lasso = Lasso(alpha=0.001)  # You can also tune this alpha
lasso.fit(X_check, y_check)
y_pred_lasso = lasso.predict(X_check)
lasso_r2 = r2_score(y_check, y_pred_lasso)
print("Lasso Regression R² Score:", lasso_r2)

# Optional: Cross-validation (5-fold)
ridge_cv_score = cross_val_score(Ridge(alpha=0.001), X_check, y_check, cv=5, scoring='r2').mean()
lasso_cv_score = cross_val_score(Lasso(alpha=0.001), X_check, y_check, cv=5, scoring='r2').mean()
print("Ridge CV R² Score:", ridge_cv_score)
print("Lasso CV R² Score:", lasso_cv_score)


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_ridge.py:215: LinAlgWarning: Ill-conditioned matrix (rcond=2.25021e-17): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T


Ridge Regression R² Score: 0.7518096807859095
Lasso Regression R² Score: 0.7155805037829743


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_ridge.py:215: LinAlgWarning: Ill-conditioned matrix (rcond=2.81491e-17): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T
/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_ridge.py:215: LinAlgWarning: Ill-conditioned matrix (rcond=2.80617e-17): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T
/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_ridge.py:215: LinAlgWarning: Ill-conditioned matrix (rcond=2.81501e-17): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T
/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_ridge.py:215: LinAlgWarning: Ill-conditioned matrix (rcond=2.82257e-17): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T
/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_ridge.py:215: LinAlgWarnin

Ridge CV R² Score: 0.7517705929203076
Lasso CV R² Score: 0.7155498512433018


In [ ]:
ols_mse = sigma_squared # Example placeholder, set this to your actual OLS model's MSE

# Compute Mallow's Cp for Ridge
n = X_check.shape[0]
p = np.sum(ridge.coef_ != 0) + 1  # number of non-zero coefficients + intercept
rss_ridge = np.sum((y_check - y_pred_ridge) ** 2)

cp_ridge = (rss_ridge / ols_mse) - n + 2 * p
print("Mallow's Cp for Ridge:", cp_ridge)

Mallow's Cp for Ridge: 18.00000056641875
